openpype/pipeline/workfile/workfile_template_builder.py

```diff
@@ -1079,15 +1117,49 @@ def set_plugin_shared_populate_data(self, key, value):
         plugin_data[key] = value
         self.builder.set_shared_populate_data(self.identifier, plugin_data)
 
+    def register_on_finished_callback(
+            self, placeholder, callback, order=None
+    ):
+        self.register_callback(
+            placeholder,
+            topic="template.finished",
+            callback=callback,
+            order=order
+        )
+
+    def register_on_depth_processed_callback(
+            self, placeholder, callback, order=0
+    ):
+        self.register_callback(
+            placeholder,
+            topic="template.depth_processed",
+            callback=callback,
+            order=order
+        )
+
+    def register_callback(self, placeholder, topic, callback, order=None):
+
+        if order is None:
+            # Match placeholder order by default
+            order = placeholder.order
+
+        # We must persist the callback over time otherwise it will be removed
+        # by the event system as a valid function reference. We do that here
+        # always just so it's easier to develop plugins where callbacks might
+        # be partials or lambdas
+        placeholder.data.setdefault("callbacks", []).append(callback)
```

Comment:
This must be handled in a different way. It is job of the object/logic which registered the callback to keep it in memory.

Reply:
I see - it does complicate the average call.

For example this https://github.com/ynput/OpenPype/blob/ed7968ec08b71a9ab732ba36f86ad61e3c2029ff/openpype/hosts/maya/plugins/template_loaders/assign_look_placeholder.py#L82-L89

And this: 
https://github.com/ynput/OpenPype/blob/d00d2d2d484ccbc0e83786c2385cea032e9ff21c/openpype/hosts/maya/plugins/template_loaders/script_placeholder.py#L155-L175

Would then also have to store each partial into a `self._callbacks` list or alike. So before each call to those `register` it should thus have:

```python
self._callbacks.append(delete_callback)
```

Is that how you'd like to see it instead?

Reply:
I call also make it so that it doesn't store it on the placeholder but instead stores it on the plug-in itself.

So this becomes:
```python
    def register_callback(self, placeholder, topic, callback, order=None):

        if order is None:
            # Match placeholder order by default
            order = placeholder.order

        # We must persist the callback over time otherwise it will be removed
        # by the event system as a valid function reference. We do that here
        # always just so it's easier to develop plugins where callbacks might
        # be partials or lambdas
        self._callbacks.append(callback)

        self.log.debug("Registering '%s' callback: %s", topic, callback)
        self.builder.add_event_callback(topic, callback, order=order)
```